### PR TITLE
LXD Profile show-machine cmd

### PR DESCRIPTION
**Pull request by SimonRichardson**
## Description of change

The following displays the LXD profiles for the machines in the
output for `show-machine`.

It assumes that the serialised name of the lxd profile is rebuildable 
on the application status side. There could be times when this 
might not happen, but later commits should improve the reliability 
of this. 

## QA steps

It assumes that you've got the feature flag enabled, along with a 
charm with a LXD profile.

```sh
juju bootstrap localhost test
juju deploy ./testcharms/charm-repo/quantal/lxd-profile
juju show-machine
```

## Documentation changes

Changes the output for `show-machine` to to show the LXD profile
for a machine. 

## Bug reference

N/A

## Example

This is the output for `show-machine`:

```
$ juju show-machine
model: default
machines:
  "0":
    juju-status:
      current: pending
      since: 11 Oct 2018 16:40:16+01:00
    dns-name: 10.88.71.92
    ip-addresses:
    - 10.88.71.92
    instance-id: juju-dac318-0
    machine-status:
      current: running
      message: Running
      since: 11 Oct 2018 16:40:23+01:00
    series: bionic
    hardware: arch=amd64 cores=0 mem=0M
    lxd-profiles:
      juju-default-lxd-profile-0:
        config:
          environment.http_proxy: ""
          linux.kernel_modules: openvswitch,nbd,ip_tables,ip6_tables
          security.nesting: "true"
          security.privileged: "true"
        description: lxd profile for testing, black list items grouped commented out
        devices:
          bdisk:
            source: /dev/loop0
            type: unix-block
          sony:
            productid: 51da
            type: usb
            vendorid: 0fce
          tun:
            path: /dev/net/tun
            type: unix-char
```

This is the output for `status --format=yaml`:

```sh
$ juju status --format=yaml
model:
  name: default
  type: iaas
  controller: xxx
  cloud: localhost
  region: localhost
  version: 2.5-beta1.1
  model-status:
    current: available
    since: 11 Oct 2018 16:40:02+01:00
  sla: unsupported
machines:
  "0":
    juju-status:
      current: started
      since: 11 Oct 2018 16:41:08+01:00
      version: 2.5-beta1.1
    dns-name: 10.88.71.92
    ip-addresses:
    - 10.88.71.92
    instance-id: juju-dac318-0
    machine-status:
      current: running
      message: Running
      since: 11 Oct 2018 16:40:23+01:00
    series: bionic
    network-interfaces:
      eth0:
        ip-addresses:
        - 10.88.71.92
        mac-address: 00:16:3e:5e:1f:50
        gateway: 10.88.71.1
        is-up: true
    hardware: arch=amd64 cores=0 mem=0M
    lxd-profiles:
      juju-default-lxd-profile-0:
        config:
          environment.http_proxy: ""
          linux.kernel_modules: openvswitch,nbd,ip_tables,ip6_tables
          security.nesting: "true"
          security.privileged: "true"
        description: lxd profile for testing, black list items grouped commented out
        devices:
          bdisk:
            source: /dev/loop0
            type: unix-block
          sony:
            productid: 51da
            type: usb
            vendorid: 0fce
          tun:
            path: /dev/net/tun
            type: unix-char
applications:
  lxd-profile:
    charm: local:bionic/lxd-profile-0
    series: bionic
    os: ubuntu
    charm-origin: local
    charm-name: lxd-profile
    charm-rev: 0
    exposed: false
    application-status:
      current: active
      message: ready
      since: 11 Oct 2018 16:41:12+01:00
    units:
      lxd-profile/0:
        workload-status:
          current: active
          message: ready
          since: 11 Oct 2018 16:41:12+01:00
        juju-status:
          current: idle
          since: 11 Oct 2018 16:41:12+01:00
          version: 2.5-beta1.1
        leader: true
        machine: "0"
        public-address: 10.88.71.92
    version: "18.04"
    endpoint-bindings:
      another: ""
      ubuntu: ""
controller:
  timestamp: 16:41:37+01:00
```

### CODE REVIEW

**Comment by SimonRichardson** [comment]
https://trello.com/c/B5tpVrWy/497-update-juju-show-machine-to-show-profile-details-active-on-the-machine

**Comment by SimonRichardson** [comment]
!!build!!

**Comment by SimonRichardson** [comment]
Moved the acceptance test to #9313, so that we can view it with removal of the feature flag.

File: acceptancetests/repository/charms/lxd-profile/README [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# Overview
```
Comment (hmlanigan) [inline acceptancetests/repository/charms/lxd-profile/README]: Should these files, acceptancetests/*, go in the acceptance test pr?

File: core/lxdprofile/name.go [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package lxdprofile
+
+import (
+	"fmt"
+)
+
+// PrefixName defines what the prefix for the name will be
+const PrefixName = "juju"
```
Comment (hmlanigan) [inline core/lxdprofile/name.go]: names.Juju instead, from juju/juju/names/names.go?

File: apiserver/facades/client/client/status.go [hunk 3]
```diff
@@ -885,6 +927,25 @@ func (c *statusContext) makeMachineStatus(machine *state.Machine) (status params
 		status.Hardware = hc.String()
 	}
 	status.Containers = make(map[string]params.MachineStatus)
+
+	lxdProfiles := make(map[string]params.LXDProfile)
+	charmProfiles, err := machine.CharmProfiles()
+	if err == nil {
+		for _, v := range charmProfiles {
+			// how do we get the charmProfile
```
Comment (hmlanigan) [inline apiserver/facades/client/client/status.go]: Is this comment still relevant?

File: apiserver/facades/client/client/status.go [hunk 4]
```diff
@@ -621,19 +639,30 @@ func fetchAllApplicationsAndUnits(
 		allBindingsByApp[bindings.AppName] = bindings.Bindings
 	}
 
+	charmProfileNames := make(map[charm.URL]lxdProfileName)
 	for _, app := range applications {
 		appMap[app.Name()] = app
 		appUnits := allUnitsByApp[app.Name()]
+		charmURL, _ := app.CharmURL()
```
Comment (hmlanigan) [inline apiserver/facades/client/client/status.go]: you should be able to get the profile from app.Charm().LXDProfile() also?  Not sure if that simplifies things or not?

File: apiserver/facades/client/client/status.go [hunk 5]
```diff
@@ -621,19 +639,30 @@ func fetchAllApplicationsAndUnits(
...
 			// the latest store revision can be looked up.
-			charmURL, _ := app.CharmURL()
 			if charmURL.Schema == "cs" {
 				latestCharms[*charmURL.WithRevision(-1)] = nil
 			}
 		}
+
+		// Create some lazy mapping between charm urls and serialised lxd
+		// profile names
+		charmProfileNames[*charmURL] = lxdProfileName{
```
Comment (hmlanigan) [inline apiserver/facades/client/client/status.go]: The machines instanceData should have a slide of profile names used by the individual machines. machine.CharmProfiles()

File: core/lxdprofile/name.go (hunk 2 above)
Comment (SimonRichardson) [reply core/lxdprofile/name.go]: Yeah, np.

File: apiserver/facades/client/client/status.go (hunk 3 above)
Comment (SimonRichardson) [reply apiserver/facades/client/client/status.go]: nope.

File: apiserver/facades/client/client/status.go (hunk 4 above)
Comment (SimonRichardson) [reply apiserver/facades/client/client/status.go]: Potentially.

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**